README
NAME
    DateTimeX::Factory - DateTime factory module with default timezone.

VERSION
    This document describes DateTimeX::Factory version 0.03.

SYNOPSIS
        use DateTimeX::Factory;

        #Object interface
        my $factory = DateTimeX::Factory->new(
            time_zone => 'Asia/Tokyo',
        );
        my $now = $factory->now;

        #Class interface
        DateTimeX::Factory->set_time_zone(DateTime::TimeZone->new(name => 'Asia/Tokyo'));
        my $now = DateTimeX::Factory->now;

DESCRIPTION
    DateTime factory module with default timezone. This module include
    wrapper of default constructors and some useful methods.

METHODS
  "set_time_zone($time_zone)"
    If called as instance method, set time_zone for its factory instance
    methods. If called as class method, set time_zone for class methods.

        #Object interface
        {
            my $factory = DateTimeX::Factory->new();
            say $factory->now->time_zone->name; # floating

            $factory->set_time_zone(DateTime::TimeZone->new(name => 'Asia/Tokyo'));
            say $factory->now->time_zone->name; # Asia/Tokyo

            #This is also OK
            $factory->set_time_zone('Asia/Tokyo');
            say $factory->now->time_zone->name; # Asia/Tokyo

            # set to default time zone (floating)
            $factory->set_time_zone;
        }

        #Class interface
        {
            DateTimeX::Factory->set_time_zone(DateTime::TimeZone->new(name => 'Asia/Tokyo'));
            say DateTimeX::Factory->now->time_zone->name; #Asia/Tokyo

            #This is also OK
            DateTimeX::Factory->set_time_zone('Asia/Tokyo');
            say DateTimeX::Factory->now->time_zone->name; #Asia/Tokyo

            # set to default time zone (floating)
            DateTimeX::Factory->set_time_zone;
        }

  "get_time_zone($time_zone)"
    Get DateTime::TimeZone instance of current time zone.

  "create(%params)"
    Call DateTime->new with default parameter.

      my $datetime = DateTimeX::Factory->create(year => 2012, month => 1, day => 24, hour => 23, minute => 16, second => 5);

  "now(%params)", "today(%params)", "from_epoch(%params)", "last_day_of_month(%params)", "from_day_of_year(%params)"
    See document of DateTime. But, these methods create DateTime instance by
    original method with default parameter.

  "strptime($string, $pattern)"
    Parse string by DateTime::Format::Strptime with default parameter.

      my $datetime = DateTimeX::Factory->strptime('2012-01-24 23:16:05', '%Y-%m-%d %H:%M:%S');

  "from_mysql_datetime($string)"
    Parse MySQL DATETIME string with default parameter.

      #equals my $datetime = DateTimeX::Factory->strptime('2012-01-24 23:16:05', '%Y-%m-%d %H:%M:%S');
      my $datetime = DateTimeX::Factory->from_mysql_datetime('2012-01-24 23:16:05');

  "from_mysql_date($string)"
    Parse MySQL DATE string with default parameter.

      #equals my $date = DateTimeX::Factory->strptime('2012-01-24', '%Y-%m-%d');
      my $date = DateTimeX::Factory->from_mysql_date('2012-01-24');

  "from_ymd($string, $delimiter)"
    Parse string like DateTime::ymd return value with default parameter.

      #equals my $date = DateTimeX::Factory->strptime('2012/01/24', '%Y/%m/%d');
      my $date = DateTimeX::Factory->from_ymd('2012-01-24', '/');

  "tommorow(%params)"
    Create next day DateTime instance.

      #equals my $tommorow = DateTimeX::Factory->today->add(days => 1);
      my $tommorow = DateTimeX::Factory->tommorow;

  "yesterday(%params)"
    Create previous day DateTime instance.

      #equals my $yesterday = DateTimeX::Factory->today->subtract(days => 1);
      my $yesterday = DateTimeX::Factory->yesterday;

DEPENDENCIES
    Perl 5.10.0 or later. Data::Validator DateTime DateTime::Format::MySQL
    DateTime::Format::Strptime DateTime::TimeZone Mouse

BUGS
    All complex software has bugs lurking in it, and this module is no
    exception. If you find a bug please either email me, or add the bug to
    cpan-RT.

SEE ALSO
    perl

AUTHOR
    Nishibayashi Takuji <takuji@senchan.jp>

LICENSE AND COPYRIGHT
    Copyright (c) 2012, Nishibayashi Takuji. All rights reserved.

    This library is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself.



Changes
Revision history for Perl extension DateTimeX::Factory

{{$NEXT}}

1.00 2013-05-14T15:05:28Z

    - no codes and documents are changed (version number issue fix)

0.10.0 2013-05-12T16:13:22Z

    - all constructor options thrown to DateTime factory methods
    - remove Mouse and Data::Validator
    - change Changes format
    [IMCOMPATIBLE CHANGES]
    - remove DateTimeX::Factory::Declare

v0.03 2012-02-01T12:11:34Z
    - fix pod spell miss
    - add set_time_zone and get_time_zone method
    - add DateTimeX::Factory::Declare

v0.02 2012-01-29T11:44:47Z
    - fix dependencies (No code changes)

v0.01 2012-01-25T23:34:30Z
    - original version


Build.PL
# =========================================================================
# THIS FILE IS AUTOMATICALLY GENERATED BY MINILLA.
# DO NOT EDIT DIRECTLY.
# =========================================================================

use 5.008_001;

use strict;
use warnings;
use utf8;

use Module::Build;
use File::Basename;
use File::Spec;
use CPAN::Meta;
use CPAN::Meta::Prereqs;

my %args = (
    license              => 'perl',
    dynamic_config       => 0,

    configure_requires => {
        'Module::Build' => 0.38,
    },

    name            => 'DateTimeX-Factory',
    module_name     => 'DateTimeX::Factory',
    allow_pure_perl => 0,

    script_files => [glob('script/*'), glob('bin/*')],

    test_files           => ((-d '.git' || $ENV{RELEASE_TESTING}) && -d 'xt') ? 't/ xt/' : 't/',
    recursive_test_files => 1,
);
if (-d 'share') {
    $args{share_dir} = 'share';
}

my $builder = Module::Build->subclass(
    class => 'MyBuilder',
    code => q{
        sub ACTION_installdeps {
            die "Do not run installdeps. Run `cpanm --installdeps .` instead.\n";
        }
    }
)->new(%args);
$builder->create_build_script();

my $mbmeta = CPAN::Meta->load_file('MYMETA.json');
my $meta = CPAN::Meta->load_file('META.json');
my $prereqs_hash = CPAN::Meta::Prereqs->new(
    $meta->prereqs
)->with_merged_prereqs(
    CPAN::Meta::Prereqs->new($mbmeta->prereqs)
)->as_string_hash;
my $mymeta = CPAN::Meta->new(
    {
        %{$meta->as_struct},
        prereqs => $prereqs_hash
    }
);
print "Merging cpanfile prereqs to MYMETA.yml\n";
$mymeta->save('MYMETA.yml', { version => 1.4 });
print "Merging cpanfile prereqs to MYMETA.json\n";
$mymeta->save('MYMETA.json', { version => 2 });


lib/DateTimeX/Factory/Declare.pm
package DateTimeX::Factory::Declare;
use strict;
use warnings;
use 5.010_000;

use Data::Validator;

use DateTimeX::Factory;

{
    my $FACTORY_CACHE = {};
    sub _factory_cache {$FACTORY_CACHE}
}
{

    my @METHODS = (
        [create => 'new'],
        qw/
            from_epoch
            now
            today
            yesterday
            tommorow
            last_day_of_month
            from_day_of_year
            strptime
            from_mysql_datetime
            from_mysql_date
            from_ymd
        /,
    );
    sub import {
        my $class  = shift;
        my $caller = caller;
        state $validator = Data::Validator->new(
            time_zone => {isa => 'DateTime::TimeZone', coerce => 1, xor => [qw/factory/], optional => 1},
            factory   => {isa => 'DateTimeX::Factory', xor => [qw/time_zone/], optional => 1},
        );
        my ($args) = $validator->validate(@_);

        my $time_zone = $args->{time_zone} if exists $args->{time_zone};
        my $factory;
        if(exists $args->{factory}) {
            $factory = $args->{factory};
        } elsif (exists $args->{time_zone}) {
            _factory_cache->{$time_zone->name} ||= DateTimeX::Factory->new(time_zone => $time_zone);
            $factory = _factory_cache->{$time_zone->name};
        } else {
            $factory = 'DateTimeX::Factory';
        }
        for my $meth (@METHODS) {
            my $origin = ref $meth ? $meth->[0] : $meth;
            my $alias  = ref $meth ? $meth->[1] : $meth;
            my $code = sub {
                $factory->$origin(@_);
            };
            {
                no strict 'refs';
                no warnings 'redefine';
                *{"$caller\::dt_$alias"} = $code;
            }
        }
    }
}

1;
__END__

=head1 NAME

DateTimeX::Factory::Declare - DateTimeX::Factory function interface.

=head1 VERSION

This document describes DateTimeX::Factory::Declare version 0.03.

=head1 SYNOPSIS

    use DateTimeX::Factory::Declare;

    DateTimeX::Factory->set_time_zone(DateTime::TimeZone->new(name => 'Asia/Tokyo'));
    my $dt = dt_new(year => 2011, month => 2, day => 1); #call DateTimeX::Factory->create
    my $now = dt_now;
    my $today = dt_today;
    my $yesterday = dt_yesterday;
    my $someday = dt_strptime('2011-02-01', '%F');

    #Set timezone
    use DateTimeX::Factory::Declare  (time_zone => 'Asia/Tokyo');

    #Use DateTimeX::Factory instance
    use DateTimeX::Factory;
    use DateTimeX::Factory::Declare  (factory => DateTimeX::Factory->new(time_zone => 'UTC'));


=head1 DESCRIPTION

DateTimeX::Factory function interface.

Can call DateTimeX::Factory method as dt_{method_name}(%params);

=head1 FUNCTIONS

=head2 C<< dt_new(%params) >>, C<< dt_now(%params) >>, C<< dt_today(%params) >>, C<< dt_from_epoch(%params) >>, C<< dt_last_day_of_month(%params) >>, C<< dt_from_day_of_year(%params) >>

See document of L<DateTime> and L<DateTimeX::Factory>.

=head2 C<< dt_strptime($string, $pattern) >>, C<< dt_from_mysql_datetime($string) >>, C<< dt_from_mysql_date($string) >>, C<< dt_from_ymd($string, $delimiter) >>

See document of L<DateTimeX::Factory>.


=head1 SEE ALSO

L<DateTimeX::Factory>

=head1 AUTHOR

Nishibayashi Takuji E<lt>takuji@senchan.jpE<gt>

=cut


lib/DateTimeX/Factory.pm
package DateTimeX::Factory;
use 5.010_001;
use strict;
use warnings;

our $VERSION = '0.10.0';

use Class::Accessor::Lite (
    new => 0,
    rw  => [qw/default_options/],
);

use DateTime;
use DateTime::Format::Strptime;

sub new {
    my ($class, %default_options) = @_;
    bless {default_options => \%default_options}, $class;
}

sub create            {shift->_call_factory_method('new' => @_)}
sub now               {shift->_call_factory_method('now' => @_)}
sub from_epoch        {shift->_call_factory_method('from_epoch' => @_)}
sub today             {shift->_call_factory_method('today' => @_)}
sub last_day_of_month {shift->_call_factory_method('last_day_of_month' => @_)}
sub from_day_of_year  {shift->_call_factory_method('from_day_of_year' => @_)}

sub _call_factory_method {
    my ($self, $method, @params) = @_;
    DateTime->$method(%{$self->default_options}, @params);
}

sub strptime {
    my ($self, $string, $pattern) = @_;
    return DateTime::Format::Strptime->new(
        pattern => $pattern,
        %{$self->default_options},
    )->parse_datetime($string);
}

sub from_mysql_datetime {
    my ($self, $string) = @_;

    return if !defined $string ||  $string eq '0000-00-00 00:00:00';

    return $self->strptime($string, '%Y-%m-%d %H:%M:%S');
}

sub from_ymd {
    my ($self, $string, $delimiter) = @_;

    $delimiter //= '-';
    return $self->strptime($string, join($delimiter, '%Y','%m','%d'));
}

sub from_mysql_date {
    my ($self, $string) = @_;

    return if !defined $string ||  $string eq '0000-00-00';
    return $self->from_ymd($string);
}

sub yesterday {shift->today(@_)->subtract(days => 1)}
sub tommorow  {shift->today(@_)->add(days => 1)}

1;
__END__

=head1 NAME

DateTimeX::Factory - DateTime factory module with default options.

=head1 VERSION

This document describes DateTimeX::Factory version 0.10.0.

=head1 SYNOPSIS

    use DateTimeX::Factory;

    my $factory = DateTimeX::Factory->new(
        time_zone => 'Asia/Tokyo',
    );
    my $now = $factory->now;

=head1 DESCRIPTION

DateTime factory module with default options.
This module include wrapper of default constructors and some useful methods.

=head1 METHODS

=head2 C<< new(%params) >>

Create factory instance. all parameters thrown to factory methods

  my $factory = DateTimeX::Factory->new(time_zone => 'Asia/Tokyo');

=head2 C<< create(%params) >>

Call DateTime->new with default parameter.

  my $datetime = $factory->create(year => 2012, month => 1, day => 24, hour => 23, minute => 16, second => 5);

=head2 C<< now(%params) >>, C<< today(%params) >>, C<< from_epoch(%params) >>, C<< last_day_of_month(%params) >>, C<< from_day_of_year(%params) >>

See document of L<DateTime>.
But, these methods create DateTime instance by original method with default parameter.

=head2 C<< strptime($string, $pattern) >>

Parse string by DateTime::Format::Strptime with default parameter.

  my $datetime = $factory->strptime('2012-01-24 23:16:05', '%Y-%m-%d %H:%M:%S');

=head2 C<< from_mysql_datetime($string) >>

Parse MySQL DATETIME string with default parameter.

  #equals my $datetime = $factory->strptime('2012-01-24 23:16:05', '%Y-%m-%d %H:%M:%S');
  my $datetime = $factory->from_mysql_datetime('2012-01-24 23:16:05');

=head2 C<< from_mysql_date($string) >>

Parse MySQL DATE string with default parameter.

  #equals my $date = $factory->strptime('2012-01-24', '%Y-%m-%d');
  my $date = $factory->from_mysql_date('2012-01-24');

=head2 C<< from_ymd($string, $delimiter) >>

Parse string like DateTime::ymd return value with default parameter.

  #equals my $date = $factory->strptime('2012/01/24', '%Y/%m/%d');
  my $date = $factory->from_ymd('2012-01-24', '/');

=head2 C<< tommorow(%params) >>

Create next day DateTime instance.

  #equals my $tommorow = $factory->today->add(days => 1);
  my $tommorow = $factory->tommorow;

=head2 C<< yesterday(%params) >>

Create previous day DateTime instance.

  #equals my $yesterday = $factory->today->subtract(days => 1);
  my $yesterday = $factory->yesterday;

=head1 DEPENDENCIES

Perl 5.10.1 or later.
L<Class::Accessor::Lite>
L<DateTime>
L<DateTime::Format::Strptime>

=head1 BUGS

All complex software has bugs lurking in it, and this module is no
exception. If you find a bug please either email me, or add the bug
to cpan-RT.

=head1 SEE ALSO

L<perl>

=head1 AUTHOR

Nishibayashi Takuji E<lt>takuji31@gmail.comE<gt>

=head1 LICENSE AND COPYRIGHT

Copyright (c) 2012, Nishibayashi Takuji. All rights reserved.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=cut
